### PR TITLE
cmake: Support multi-configuration generator

### PR DESCRIPTION
**Pull request by jcfr**
When using IDE like Visual Studio or XCode, the build type is not
known at configuration time, the developer usually select it when
building the project.

### CODE REVIEW

**Comment by jcfr** [comment]
Cc: @phcerdan

File: CMakeLists.txt [hunk 1]
```diff
@@ -14,17 +14,22 @@ endif()
...
 set(Eigen_BINARY_DIR ${Eigen3_BINARY_DIR})
 
-# guard against bad build-type strings
-
-if (NOT CMAKE_BUILD_TYPE)
-  set(CMAKE_BUILD_TYPE "Release")
+# Set a default build type if none was specified
+if(NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
+  message(STATUS "Setting build type to 'Release' as none was specified.")
+  set(CMAKE_BUILD_TYPE Release CACHE STRING "Choose the type of build." FORCE)
```
Comment (ggael) [inline CMakeLists.txt]: CMAKE_BUILD_TYPE is natively handled by cmake, not need to add 

    CACHE STRING "Choose the type of build." FORCE

nor 

    set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS "Debug" "Release" "MinSizeRel" "RelWithDebInfo")


Maybe CMAKE_BUILD_TYPE should just be ignored if cmake uses CMAKE_CONFIGURATION_TYPES?

**Comment by ggael** [comment]
If you can, please better send your PR there: https://bitbucket.org/eigen/eigen

File: CMakeLists.txt (hunk 1 above)
Comment (jcfr) [reply CMakeLists.txt]: This is the standard practice to deal with multi-configuration IDE and initialized CMAKE_BUILD_TYPE.

By default, it turns out that `CMAKE_BUILD_TYPE` has no value

**Comment by jcfr** [comment]
Closing. PR submitted to upstream as https://bitbucket.org/eigen/eigen/pull-requests/482/cmake-support-multi-configuration

**Pull request closed without merging**